### PR TITLE
Mithril 2 update

### PR DESCRIPTION
**Pull request by askvortsov1**

### CODE REVIEW

File: js/src/forum/components/PostQuoteButton.js [hunk 1]
```diff
@@ -1,27 +1,29 @@
-import Button from 'flarum/components/Button';
-import extract from 'flarum/utils/extract';
+import Fragment from 'flarum/Fragment';
+import icon from 'flarum/helpers/icon';
 
 import reply from '../utils/reply';
 
-export default class PostQuoteButton extends Button {
+export default class PostQuoteButton extends Fragment {
```
Comment (clarkwinkelmann) [inline js/src/forum/components/PostQuoteButton.js:6]: `PostQuoteButton` should be moved to the `fragments` folder like `AutocompleteDropdown`

Comment (askvortsov1) [reply js/src/forum/components/PostQuoteButton.js]: Fixed!

**Review by franzliedke — CHANGES_REQUESTED** [review]
I will have a few more small questions tomorrow, but I need to go to bed now...

File: js/src/forum/addPostMentionPreviews.js [hunk 2]
```diff
@@ -4,18 +4,19 @@ import PostPreview from 'flarum/components/PostPreview';
 import LoadingIndicator from 'flarum/components/LoadingIndicator';
 
 export default function addPostMentionPreviews() {
-  extend(CommentPost.prototype, 'config', function() {
-    const contentHtml = this.props.post.contentHtml();
+  extend(CommentPost.prototype, 'onupdate', function() {
```
Comment (franzliedke) [inline js/src/forum/addPostMentionPreviews.js:7]: The way I see it, this entire block was previously executed on first render, now it won't be. :thinking:

File: js/src/forum/fragments/AutocompleteDropdown.js [hunk 3]
```diff
@@ -24,8 +23,10 @@ export default class AutocompleteDropdown extends Component {
   }
 
   hide() {
-    this.$().hide();
-    this.active = false;
+    if (this.element) {
+      this.$().hide();
+      this.active = false;
+    }
```
Comment (franzliedke) [inline js/src/forum/fragments/AutocompleteDropdown.js:29]: What scenario does this prevent?

Comment (askvortsov1) [reply js/src/forum/fragments/AutocompleteDropdown.js:29]: At times when testing, hide() was called when the element DOM did not exist, so it errored.

File: js/src/forum/fragments/PostQuoteButton.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        {icon('fas fa-quote-left', { className: 'Button-icon' })}
+        {app.translator.trans('flarum-mentions.forum.post.quote_button')}
+      </button>
+    );
+  }
+
+  oncreate(vnode) {
+    super.oncreate(vnode);
+
+    $(document).on('mousedown', this.hide.bind(this));
```
Comment (franzliedke) [inline js/src/forum/fragments/PostQuoteButton.js:28]: We should unbind this somehow when destroying the component.

That seems to be a general problem with "fragments"... :thinking:

File: js/src/forum/fragments/PostQuoteButton.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
+import Fragment from 'flarum/Fragment';
+import icon from 'flarum/helpers/icon';
+
+import reply from '../utils/reply';
+
+export default class PostQuoteButton extends Fragment {
+  constructor(post) {
+    super();
```
Comment (franzliedke) [inline js/src/forum/fragments/PostQuoteButton.js:8]: Doesn't seem to be needed.

Comment (askvortsov1) [reply js/src/forum/fragments/PostQuoteButton.js:8]: Errors without it. https://stackoverflow.com/a/31079103

File: js/src/forum/addPostQuoteButton.js [hunk 6]
```diff
@@ -32,6 +34,8 @@ export default function addPostQuoteButton() {
             const lastRect = rects[rects.length - 1];
             button.showEnd(lastRect.right, lastRect.bottom);
           }
+        } else {
+          button.hide();
```
Comment (franzliedke) [inline js/src/forum/addPostQuoteButton.js:38]: Isn't this already done by the onmousedown handler in the fragment itself?

File: js/src/forum/fragments/PostQuoteButton.js (hunk 4 above)
Comment (askvortsov1) [reply js/src/forum/fragments/PostQuoteButton.js:28]: We can still use onremove

File: js/src/forum/addPostQuoteButton.js (hunk 6 above)
Comment (askvortsov1) [reply js/src/forum/addPostQuoteButton.js:38]: I believe so... But would we want to remove it in this PR?

File: js/src/forum/fragments/AutocompleteDropdown.js (hunk 3 above)
Comment (clarkwinkelmann) [reply js/src/forum/fragments/AutocompleteDropdown.js:29]: jQuery calls shouldn't error if the element doesn't exist, no? Calling `.hide()` on an empty set should work just fine.

Comment (askvortsov1) [reply js/src/forum/fragments/AutocompleteDropdown.js:29]: I'll test more

File: js/src/forum/addPostQuoteButton.js (hunk 6 above)
Comment (franzliedke) [reply js/src/forum/addPostQuoteButton.js:38]: It was *added* in this PR.

Comment (askvortsov1) [reply js/src/forum/addPostQuoteButton.js:38]: Lol I didn't notice that. Removed

File: js/src/forum/fragments/PostQuoteButton.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    this.post = post;
+  }
+
+  view() {
+    return (
+      <button class="Button PostQuoteButton" onclick={() => {
+        this.hide();
+        reply(this.post, this.content);
+      }} mousedown={(e) => e.stopPropagation()}>
```
Comment (franzliedke) [inline js/src/forum/fragments/PostQuoteButton.js:18]: This would need to be `mouseup` now as well.

Comment (askvortsov1) [reply js/src/forum/fragments/PostQuoteButton.js]: Even better, that code doesn't actually run, so we should be good just removing it.

Comment (franzliedke) [reply js/src/forum/fragments/PostQuoteButton.js]: Should it have been **on**mousedown?

Comment (askvortsov1) [reply js/src/forum/fragments/PostQuoteButton.js]: Oh no! :facepalm: Yeah, that would explain the related bugs we saw, and why the mousedown handler elsewhere needed to be changed to mouseup. That being said, the way it is now works, and is a bit simpler IMO (it'll ALWAYS hide the button on mouseup, and if we clicked on it in the process, it'll also execute a new reply

**Review by franzliedke — APPROVED** [review]

**Pull request merged**